pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/JsonAsyncHttpPinotClientTransport.java

```diff
@@ -134,11 +137,23 @@ public Future<BrokerResponse> executeQueryAsync(String brokerAddress, String que
       if (_headers != null) {
         _headers.forEach((k, v) -> requestBuilder.addHeader(k, v));
       }
+      LOGGER.debug("Sending query {} to {}", query, url);
+      return requestBuilder.addHeader("Content-Type", "application/json; charset=utf-8").setBody(json.toString())
+              .execute().toCompletableFuture().thenApply(httpResponse -> {
+        LOGGER.debug("Completed query, HTTP status is {}", httpResponse.getStatusCode());
+
+        if (httpResponse.getStatusCode() != 200) {
+          throw new PinotClientException(
+                  "Pinot returned HTTP status " + httpResponse.getStatusCode() + ", expected 200");
+        }
 
-      Future<Response> response =
-          requestBuilder.addHeader("Content-Type", "application/json; charset=utf-8").setBody(json.toString())
-              .execute();
-      return new BrokerResponseFuture(response, query, url, _brokerReadTimeout);
+        String responseBody = httpResponse.getResponseBody(StandardCharsets.UTF_8);
+        try {
+          return BrokerResponse.fromJson(OBJECT_READER.readTree(responseBody));
+        } catch (JsonProcessingException e) {
+          throw new CompletionException(e);
```

Comment:
Is this correct exception to throw. Earlier we were throwing `ExecutionException` in this scenario.

Reply:
@KKcorps The completable future will wrap exceptions other than CompletionException.

In the past there would have been ExecutionException wrapping JsonProcessingException.

Now, we'll have CompletionException wrapping JsonProcessingException, however if I use CompletionException explicitly, we'll end up with CompletionException wrapping ExecutionException wrapping JsonProcessingException.

I thought avoiding one level of wrapping would be preferred since there will be a change in the hierarchy regardless.

Please advise.

Reply:
Your thought process is correct but I was thinking from backward compatibility perspective. e.g. Some user/lib was handling ExecutionException but after the upgrade, their handling block never gets triggered.